src/stm32f1/adc.cpp

```diff
@@ -12,149 +12,270 @@
+[[maybe_unused]] static constexpr auto fourth_conversion =
+  hal::bit_mask::from(15, 19);
+
+/// Fifth channel conversion in regular sequence
+[[maybe_unused]] static constexpr auto fifth_conversion =
+  hal::bit_mask::from(20, 24);
+
+/// Sixth channel conversion in regular sequence
+[[maybe_unused]] static constexpr auto sixth_conversion =
+  hal::bit_mask::from(25, 29);
+};  // namespace adc_regular_sequence_register_3
+
+/// Namespace containing the bit_mask objects that are use to manipulate the
+/// stm32f1 ADC Regular Data register
+namespace adc_regular_data_register {
+/// These bits are read only. They contain the conversion result from the
+/// regular channels. The data is left or right-aligned depending on bit 11 in
+/// ADC_CR2.
+static constexpr auto regular_data = hal::bit_mask::from(0, 15);
+
+/// In ADC1: In dual mode, these bits contain the regular data of ADC2. Refer to
+/// Section 11.9: Dual ADC mode.
+/// In ADC2 and ADC3: these bits are not used.
+[[maybe_unused]] static constexpr auto dual_mode_data =
+  hal::bit_mask::from(16, 31);
+};  // namespace adc_regular_data_register
+
+constexpr std::uintptr_t stm_apb2_base = 0x40000000UL;
+constexpr std::uintptr_t stm_adc_addr = stm_apb2_base + 0x12400;
+inline auto* adc_reg = reinterpret_cast<adc_reg_t*>(stm_adc_addr);
+
+void setup(adc::pins const& p_pin)
+{
+  auto const adc_frequency = frequency(hal::stm32f1::peripheral::adc1);
+  if (adc_frequency > 14.0_MHz) {
+    hal::safe_throw(hal::operation_not_supported(&adc_frequency));
   }
 
   // Power on adc clock
   power_on(peripheral::adc1);
 
-  // Set specified channel to analog input mode
-  configure_pin({ .port = p_channel.port, .pin = p_channel.pin }, input_analog);
+  // Derive port and pin from enum
+  hal::u8 port, pin;
+  if (hal::value(p_pin) <= 7) {
+    port = 'A';
+    pin = hal::value(p_pin);
+  } else if (hal::value(p_pin) <= 9) {
+    port = 'B';
+    pin = hal::value(p_pin) - 8;
+  } else {
+    port = 'C';
+    pin = hal::value(p_pin) - 10;
+  }
 
-  // Power on adc
-  hal::bit_modify(adc_reg->control_2)
-    .set<adc_control_register_2::ad_converter_on>();
+  // Set specified pin to analog input mode
+  configure_pin({ .port = port, .pin = pin }, input_analog);
 
-  // Set the specified channel to be sampled
-  hal::bit_modify(adc_reg->regular_sequence_3)
-    .insert<adc_regular_sequence_register_3::first_conversion>(p_channel.index);
+  // Turns on and calibrates ADC only if its first time power-up
+  if (bit_extract<adc_control_register_2::ad_converter_on>(
+        adc_reg->control_2) == 0) {
+    // Power on adc
+    hal::bit_modify(adc_reg->control_2)
+      .set<adc_control_register_2::ad_converter_on>();
 
-  // Start adc calibration
-  hal::bit_modify(adc_reg->control_2)
-    .set<adc_control_register_2::ad_calibration>();
+    // Start adc calibration
+    hal::bit_modify(adc_reg->control_2)
+      .set<adc_control_register_2::ad_calibration>();
```

Comment:
Can you explain why this modification must happen separately from the ADC converter ON set?

Reply:
Reference manual states that the ADC must have been in power-on state for at least two ADC clock cycles before starting calibration.